ext/financialacls/tests/phpunit/Civi/Financialacls/ContributionTest.php

```diff
@@ -0,0 +1,131 @@
+<?php
+/*
+ +--------------------------------------------------------------------+
+ | Copyright CiviCRM LLC. All rights reserved.                        |
+ |                                                                    |
+ | This work is published under the GNU AGPLv3 license with some      |
+ | permitted exceptions and without any warranty. For full license    |
+ | and copyright information, see https://civicrm.org/licensing       |
+ +--------------------------------------------------------------------+
+ */
+namespace Civi\Financialacls;
+
+require_once 'BaseTestClass.php';
+
+/**
+ * Class CRM_Contribute_BAO_ContributionTest
+ * @group headless
+ */
+class ContributionTest extends BaseTestClass {
+
+  /**
+   * Test the annual query returns a correct result when multiple line items
+   * are present.
+   *
+   * @throws \Civi\Core\Exception\DBQueryException
+   */
+  public function testAnnualWithMultipleLineItems(): void {
+    $this->createContributionWithTwoLineItems();
+    \Civi::settings()->set('acl_financial_type', TRUE);
+    $this->addFinancialAclPermissions([['view', 'Donation']]);
+    $sql = \CRM_Contribute_BAO_Contribution::getAnnualQuery([$this->ids['Contact']['logged_in']]);
+    $result = \CRM_Core_DAO::executeQuery($sql);
+    $result->fetch();
+    $this->assertEquals(0, $result->N);
+
+    // It didn't find any rows cos it is restricted to only find contributions where all lines are visible.
+    \CRM_Core_DAO::executeQuery('UPDATE civicrm_line_item SET financial_type_id = ' . \CRM_Core_PseudoConstant::getKey('CRM_Contribute_BAO_Contribution', 'financial_type_id', 'Donation'));
+    $result = \CRM_Core_DAO::executeQuery($sql);
+    $result->fetch();
+    $this->assertEquals(300, $result->amount);
+    $this->assertEquals(1, $result->count);
+  }
+
+  /**
+   * Create a contribution with 2 line items.
+   *
+   * This also involves creating t
```

Comment:
Agreed - you can't get anything done properly until you've had a good cup of tea.